### PR TITLE
NLOpt Support

### PR DESCRIPTION
**Pull request by costashatz**
The code is not complete yet. I am creating this PR to discuss several things:

- Will NLOpt Integration be a part of the core library? Right now, I only use it in an example.
- How should we impose stopping criteria? At the moment, in all of the other optimization algorithms we provide via Params the stopping criteria (in all of the times is number of iterations). Maybe we should somehow homogenize the provided criteria to all of our optimizers. NLOpt provides several criteria:
  * Value/tolerance of objective function
  * Value/tolerance of individual parameters
  * Number of function evaluations (iterations)
  * Actual running time
- How should we handle the fact that NLOpt always asks for `grad` even if it is not needed? And as we have templated structs, we would need to always define `utility_and_grad` even if it makes no sense to.

What do you think?

### CODE REVIEW

**Comment by jbmouret** [comment]
> The code is not complete yet. I am creating this PR to discuss several things:
> 
> Will NLOpt Integration be a part of the core library? Right now, I only use it in an example.

I think it should be optional but part of the core library (can compile without, but not "an exercice left to the reader").

> How should we impose stopping criteria? At the moment, in all of the other optimization algorithms we provide via Params the stopping criteria (in all of the times is number of iterations). Maybe we should somehow homogenize the provided criteria to all of our optimizers. NLOpt provides several criteria:
> Value/tolerance of objective function
> Value/tolerance of individual parameters
> Number of function evaluations (iterations)
> Actual running time

Homogenizing is good, but I am not sure there is a perfect way for this. Btw, we care only about the criteria 1 & 3.

> How should we handle the fact that NLOpt always asks for grad even if it is not needed? And as we have templated structs, we would need to always define utility_and_grad even if it makes no sense to.
> What do you think?

We could use an adapter:
```
template<typename F>
struct GradAdapter {
  F& _f;
  auto utility() { return _f.utility(); }
 auto utility_and_grand() { return std::make_pair<>(f.utilty(), 0); }
};
```
What do you think?

**Comment by costashatz** [comment]
> I think it should be optional but part of the core library (can compile without, but not "an exercice left to the reader").

So, it would be in the sources and documented, but not linked/included to any parts of the core library. If the user needs it in one of its executable he has to explicitly define it in his `wscript`. I agree with this approach.

> Homogenizing is good, but I am not sure there is a perfect way for this. Btw, we care only about the criteria 1 & 3.

We can impose the same criteria to all the optimizers via Params. So, every optimizer would have `Params::optimizer_name::iters` and `Params::optimizer_name::tol` and we have to alter their implementations to check for both. This is easy to do.

> We could use an adapter:
> 
> template<typename F>
> struct GradAdapter {
>   F& _f;
>   auto utility() { return _f.utility(); }
>  auto utility_and_grand() { return std::make_pair<>(f.utilty(), 0); }
> };

It's not a bad idea, but I have to test it in implementation to see if I like it.

**Comment by jbmouret** [comment]
> I think it should be optional but part of the core library (can compile without, but not "an exercice left to the reader").
> So, it would be in the sources and documented, but not linked/included to any parts of the core library. If the user needs it in one of its executable he has to explicitly define it in his wscript. I agree with this approach.

I am not sure we mean the same thing here. I mean: it would be like sferes or TBB (limbo compiles with it if it is found, without if it is not found).

**Comment by costashatz** [comment]
> I am not sure we mean the same thing here. I mean: it would be like sferes or TBB (limbo compiles with it if it is found, without if it is not found).

OK.

**Comment by fedeallocati** [comment]
> We can impose the same criteria to all the optimizers via Params. So, every optimizer would have Params::optimizer_name::iters and Params::optimizer_name::tol and we have to alter their implementations to check for both. This is easy to do.

And if someone wants only iteras or only tolerance?

**Comment by costashatz** [comment]
> And if someone wants only iteras or only tolerance?

We have to create some logic on how they should be interpreted as invalid. Thus, meaning that the user does not wish to use them. E.g. negative value for iterations.

**Comment by costashatz** [comment]
> I mean: it would be like sferes or TBB (limbo compiles with it if it is found, without if it is not found).

This is done.

**Comment by fedeallocati** [comment]
I'm not sure about the define `USE_NLOPT`. I think that if the user explicitly uses NLOPT as optimizer, then he must also explicitly add the dependency in it's target. I think the same for the `USE_SFERES` define.
The only one that makes sense for me is the TBB one, because in case the user doesn't include TBB, there is an alternative way too perform the things, while in the case of NLOPT, if the user doesn't includes it, it's pointless to use it as optimizer.

**Comment by costashatz** [comment]
> I'm not sure about the define USE_NLOPT. I think that if the user explicitly uses NLOPT as optimizer, then he must also explicitly add the dependency in it's target. I think the same for the USE_SFERES define.
> The only one that makes sense for me is the TBB one, because in case the user doesn't include TBB, there is an alternative way too perform the things, while in the case of NLOPT, if the user doesn't includes it, it's pointless to use it as optimizer.

But NLOpt is not ONE optimizer, is a library of many optimizers. And so we provide a general support as part of the core library so that the user can easily use NLOpt. The other way is a little bit more complicated to the user. Although, I do not really have a strong preference. @jbmouret What do you say?

**Comment by jbmouret** [comment]
On 09 Nov 2015, at 12:58, Konstantinos Chatzilygeroudis <notifications@github.com> wrote:

> I'm not sure about the define USE_NLOPT. I think that if the user explicitly uses NLOPT as optimizer, then he must also explicitly add the dependency in it's target. I think the same for the USE_SFERES define.
> The only one that makes sense for me is the TBB one, because in case the user doesn't include TBB, there is an alternative way too perform the things, while in the case of NLOPT, if the user doesn't includes it, it's pointless to use it as optimizer.
> 
> But NLOpt is not ONE optimizer, is a library of many optimizers. And so we provide a general support as part of the core library so that the user can easily use NLOpt. The other way is a little bit more complicated to the user. Although, I do not really have a strong preference. @jbmouret What do you say?


I do not really see the issue: if the user wants to use a NlOpt optimiser, he will have to explicit add it to its script and call it in the right way. That does not mean that NlOpt optimisers are not �first class citizen�...


� Jean-Baptiste Mouret
http://pages.isir.upmc.fr/~mouret/

**Comment by costashatz** [comment]
> I do not really see the issue: if the user wants to use a NlOpt optimiser, he will have to explicit add it to its script and call it in the right way. That does not mean that NlOpt optimisers are not �first class citizen�...

The issue is that now he doesn't have to explicitly specify it. It is provided if found. If it is not found, the code using NLOpt compiles but does not work.

@fedeallocati says that it's better to let the user explicitly specify it in his `wscript` every time he wants to use it.

With the former, the `nlopt.hpp` is always included (in `opt.hpp`) and a define `USE_NLOPT` provides or excludes the usage of NLOpt structs. Any code that has NLOpt structs compiles, but does not provide any useful data. With the later (Fede's), NLOpt won't be included in the main includes and if a user wants to use it, he will have to explicitly define it in his `wscript` and include `nlopt.hpp`.

I do not have any strong preference. I like the way we have it now because the user only needs to specify NLOpt once (or even he doesn't have if he has installed it globally - `/usr/` or `/usr/local/`) and then everything works. On the other hand, I like Fede's idea because the code is cleaner in the library and if it is well documented there won't be any problem for the users to use it.

**Comment by fedeallocati** [comment]
> I do not have any strong preference. I like the way we have it now because the user only needs to specify NLOpt once (or even he doesn't have if he has installed it globally - /usr/ or /usr/local/).
That's not what I meant. What I meant is just remove the define `USE_NLOPT`. If the user uses it, then the code won't compile unless on his wscript he specifies NLOPT (the same way we specify BOOST, EIGEN, TBB, etc). So the NLOpt class (which I know is a wrapper for multiple NLOPT optimizers) will assume that the paths to NLOpt header files are defined. In this way, the code will be cleaner.

**Comment by costashatz** [comment]
> That's not what I meant. What I meant is just remove the define USE_NLOPT. If the user uses it, then the code won't compile unless on his wscript he specifies NLOPT (the same way we specify BOOST, EIGEN, TBB, etc). So the NLOpt class (which I know is a wrapper for multiple NLOPT optimizers) will assume that the paths to NLOpt header files are defined. In this way, the code will be cleaner.

Yes, but if someone does not want to use NLOpt, how will he compile it?

**Comment by fedeallocati** [comment]
I could probably be wrong, but if I remember correctly, since it's templated code, as long as one doesn't declare a concrete type for the class, it won't be compiled.

**Comment by costashatz** [comment]
> I could probably be wrong, but if I remember correctly, since it's templated code, as long as one doesn't declare a concrete type for the class, it won't be compiled.

You are right, but we have a test using it. Maybe I could add the `USE_NLOPT` there.

**Comment by fedeallocati** [comment]
>You are right, but we have a test using it. Maybe I could add the USE_NLOPT there.

In that test, we should add NLOPT to the target in the wscript

**Comment by costashatz** [comment]
>That's not what I meant. What I meant is just remove the define USE_NLOPT. If the user uses it, then the code won't compile unless on his wscript he specifies NLOPT (the same way we specify BOOST, EIGEN, TBB, etc). So the NLOpt class (which I know is a wrapper for multiple NLOPT optimizers) will assume that the paths to NLOpt header files are defined. In this way, the code will be cleaner.

It should be okay now!

File: src/limbo/init/random_sampling.hpp [hunk 1]
```diff
@@ -20,7 +18,6 @@ namespace limbo {
                     Eigen::VectorXd new_sample(F::dim_in);
                     for (int i = 0; i < F::dim_in; i++)
                         new_sample[i] = tools::rand<double>(0, 1);
-                    std::cout << "random sample:" << new_sample.transpose() << std::endl;
```
Comment (fedeallocati) [inline src/limbo/init/random_sampling.hpp]: Why has this been removed?

Comment (costashatz) [inline src/limbo/init/random_sampling.hpp]: Because in no other initialization function we output anything. We shouldn't be inconsistent. It shouldn't really be part of this PR, but I just saw it and did it.

File: src/limbo/opt/nlopt.hpp [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+#else
+#include <Eigen/Core>
+
+#include <vector>
+
+#include <nlopt.hpp>
+
+namespace limbo {
+    namespace opt {
+        template <typename Params, nlopt::algorithm Algorithm = nlopt::LD_MMA>
```
Comment (fedeallocati) [inline src/limbo/opt/nlopt.hpp]: Just curious, why default to LD_MMA?

Comment (costashatz) [inline src/limbo/opt/nlopt.hpp]: Only because it was the default in their examples. No other reason. If you have any other preference, please do not hesitate. You can see the NLOpt algorithms [here](http://ab-initio.mit.edu/wiki/index.php/NLopt_Algorithms).

**Comment by costashatz** [comment]
What's the status on this? Does anyone have strong objections or do we merge?

**Comment by fedeallocati** [comment]
I'm using this branch to compare with BayesOpt, and I think we are missing a parameter to specify the number of iterations to optimize with NlOpt (the equivalent of RProp *iters*)

**Comment by costashatz** [comment]
> I'm using this branch to compare with BayesOpt, and I think we are missing a parameter to specify the number of iterations to optimize with NlOpt (the equivalent of RProp iters)

It should be fixed.

**Comment by fedeallocati** [comment]
I think we should add a different NLOpt optimizer class for functors that don't have gradient.

**Comment by costashatz** [comment]
> I think we should add a different NLOpt optimizer class for functors that don't have gradient.

Okay. I am on to it.

**Comment by costashatz** [comment]
How will we solve the parameters for the optimizers? Should I just add only `iterations` for the moment and open up an issue? Because it is not only related to NLOpt, but also to the other optimizers.

**Comment by jbmouret** [comment]
Let's do it this way.

**Comment by costashatz** [comment]
Any other issues or do we merge?

**Comment by jbmouret** [comment]
I would like to try on m'y mac first.

**Comment by costashatz** [comment]
Just compiled the extensive tests. Everything seems fine!

**Comment by fedeallocati** [comment]
Since we say that NLOpt should be optional, shouldn't also the test "test_nlopt.hpp" be optional?

**Comment by costashatz** [comment]
> Since we say that NLOpt should be optional, shouldn't also the test "test_nlopt.hpp" be optional?

When no NLOpt is found, the test does nothing. We could add the test in the wscript, but just using the preprocessor works also (`USE_NLOPT`).

**Comment by fedeallocati** [comment]
I think that not declaring the build in the wscript if the variable USE_NLOPT is not defined should be fine

**Comment by jbmouret** [comment]
Btw, I have a patch to submit for the compilation on mac.

� JBM
On 25 Nov 2015, at 18:10, Konstantinos Chatzilygeroudis <notifications@github.com> wrote:

> Since we say that NLOpt should be optional, shouldn't also the test "test_nlopt.hpp" be optional?
> 
> When no NLOpt is found, the test does nothing. We could add the test in the wscript, but just using the preprocessor works also (USE_NLOPT).
> 
> �
> Reply to this email directly or view it on GitHub.
> 

� Jean-Baptiste Mouret
http://pages.isir.upmc.fr/~mouret/

**Comment by costashatz** [comment]
> I think that not declaring the build in the wscript if the variable USE_NLOPT is not defined should be fine

Should be OK now.

@jbmouret What was the problem in mac? Maybe the static lib? Because I am linking to the shared one, but the default (in packages) is the static one.

**Comment by jbmouret** [comment]
No big deal.
� JBM
On 25 Nov 2015, at 18:21, Konstantinos Chatzilygeroudis <notifications@github.com> wrote:

> I think that not declaring the build in the wscript if the variable USE_NLOPT is not defined should be fine
> 
> Should be OK now.
> 
> @jbmouret What was the problem in mac? Maybe the static lib? Because I am linking to the shared one, but the default (in packages) is the static one.
> 
> �
> Reply to this email directly or view it on GitHub.
> 

� Jean-Baptiste Mouret
http://pages.isir.upmc.fr/~mouret/

**Comment by jbmouret** [comment]
This should be OK now.

**Comment by costashatz** [comment]
So should we merge?

**Comment by jbmouret** [comment]
**OK for merging**

**Comment by fedeallocati** [comment]
**Ok for me**
